File: 0437_Path Sum III/PathSumIII.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    private int ans = 0;
    public int PathSum(TreeNode root, int sum) {
        if(root == null) return ans;
        
        PathSum(root, sum, new List<int>());
        return ans;
        
        
    }
    
    private void PathSum(TreeNode node, int sum, IList<int> list){
        if(node == null) return;
        list.Add(node.val);
        
        int tmpSum = 0;
        for(int i=list.Count-1;i>=0;i--){
            tmpSum += list[i];
            if(tmpSum == sum){
                ans++;
            }
        }
        
        PathSum(node.left, sum, list);
        PathSum(node.right, sum, list);
        
        list.RemoveAt(list.Count - 1);
    }
}

File: 0437_Path Sum III/PathSumIII.java
class Solution {
    private int ans = 0;
    public int pathSum(TreeNode root, int targetSum) {
        traverse(root, new ArrayList<Long>(), targetSum);
        return ans;
    }
    
    private void traverse(TreeNode node, List<Long> list, int targetSum) {
        if(node == null) return;
        list.add((long)node.val);
        long sum = 0;
        for(int i=list.size()-1;i>=0;i--) {
            sum += list.get(i);
            if(sum == targetSum) ans++;
        }
        
        traverse(node.left, list, targetSum);
        traverse(node.right, list, targetSum);
        list.remove(list.size()-1);
    }
}